keps/sig-cloud-provider/2699-add-webhook-hosting-to-ccm/README.md

```diff
@@ -0,0 +1,455 @@
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP will detail enhancing the CCM framework to support cloud provider
+specific webhooks. The intent is to make it easy to either generate a binary
+or enhance the existing CCM binary to host such webhooks. We also intend to
+allow for easily linking in "standard" webhooks needed by other SIGs which
+need to be customized for particular cloud providers.
+
+## Motivation
+
+The Cloud (Provider) Controller Manager (CCM) is the binary into which the Cloud
+Provider places all the controllers needed to make a Kubernetes cluster work
+correctly on their Cloud. There are also occasions when it makes sense for a Cloud
+Provider to want these customizations to be applied inline rather than 
+asynchronously after a change has already been applied. 
+
+Our initial example of this is from SIG Storage. These would like the functionality 
+from PVL admission controller (https://github.com/kubernetes/kubernetes/issues/52617).
+This needs to be completed for cloud provider extraction to complete. Several
+Cloud Providers have indicated that this should be done inline, especially as the
+existing deprecated solution is an inline solution.
+
+### Goals
+
+Our current in tree Cloud Providers are able to stop using the existing PVL admission 
+controller and do so using the framework.
+
+### Non-Goals
+
+We are not intending to create a general admission webhook solution. This is just
```

Comment:
What is the distinction between "general admission webhook solution" and "cloud provider specific webhooks as part of the control plane"?

Reply:
To my mind something like https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html is an attempt at a general admission webhook solution. Kubebuilder will generate a general webhook server and then you just put your business logic in the hooks provided by the defaulter or validator interface. These are generally webhooks designed to be part of the workload rather than the control plane.

We are creating a framework to house K8s specific business logic (Eg the PVL) which needs to be combined with the cloud provider interface and is part of Cloud Providers control plane for Kubernetes. This doesn't preclude the use of controller runtime as part of the solution but suggests it is on its own insufficient.

Reply:
A webhook that needs the `ControllerContext` instead of just `ClientSet`?

Reply:
Not sure I understand what the point your making about the ControllerContext is?

Reply:
Sorry, my mistake. I was wondering why we need to have the context of a controller manager for the webhook.

Reply:
We shouldn't need a controller manager (or its context) for the webhook. We are leaving open the option to colocate them in the same binary for people who want to keep their deployment simple and easily separate them into two binaries for those who don't. Personally I would recommend having two binaries but I realize that is extra work.